### ISSUE TITLE
Fix redirect to accommodate sub-directory deployment

### ISSUE THREAD

**Issue by hangler** [bcgov/ExitSurveyAdmin#98]
As requested by @robinwn, provide a solution to fixing the callback URL. This will involve an `.env` with a setting.

**Comment by hangler** [comment]
@robinwn See https://github.com/bcgov/ExitSurveyAdmin/commit/10eb3b6b092f3e5e19fc01cf5201dee08c34d62f. There is now a new setting in `ClientApp/.env`: `REACT_APP_CALLBACK_LOCATION`. This needs to be set to the path the application gets deployed to, including the `/callback` endpoint.

Let me know if this fixes the problem!

**Comment by robinwn** [comment]
@hangler  - Hi Frank - that should resolve the callback url.  There is also the 'authority' and 'silent_redirect_uri' the need to be set in the deployment environment.  When possible we like to create a deployment artifact that picks up settings in the context of where the app is deployed, eg. web.config with settings for a URI prefix of the app so it knows how to construct URLs when the app is proxied.  The dotnet core api has a mechanism for this ie appsettings.<deploy_env>.json.  However, for the front end (React, Angular, Vue etc) it looks like the only option is compiling the app with the final deploy settings.  This means a build for every deployment environment we have - minimally, DEV, TEST and PROD.  not ideal.

**Comment by hangler** [comment]
Ah, great, will factor those out as well and update when done.

Yes, I don't know of a better solution than building a different artifact for each environment, though maybe there are some floating around out there... But, building each separately is how I've always done it on other projects. At the end of the day, with these front-end apps, it's a problem of getting certain values (e.g. URLs to various APIs) into the client's browser without them hitting a back-end server to retrieve them. The cleanest way is to bake those settings into the JS file they're getting served.

**Comment by hangler** [comment]
Discussion here (_caveat lector_, I have only skimmed it): https://github.com/facebook/create-react-app/issues/2353

**Comment by robinwn** [comment]
@hangler  I had a read through both https://github.com/facebook/create-react-app/issues/2353 and [andrewmclagan/react-env](https://github.com/andrewmclagan/react-env)  (several times).  This looks like a good way to go.  I'll be able to run the same 'react-env' as shown [here](https://github.com/andrewmclagan/react-env/blob/master/examples/create-react-app/README.md) in the TFS deploy which is the logical equivalent of static content deploy or the Docker startup.

Let me know if you have any concerns about this.

**Comment by hangler** [comment]
@robinwn I do not in principle, the main question is how to implement it, which I will try to figure out now!

**Comment by robinwn** [comment]
@hangler  the [create-react-app](https://github.com/andrewmclagan/react-env/tree/master/examples/create-react-app) example looks like a good starting point

**Comment by hangler** [comment]
Well that was easy 😄 I have it working locally. I added a bunch of environment variables to `.env`.; see https://github.com/bcgov/ExitSurveyAdmin/commit/f91cd2c8bdff626321a81a4fa36b0f04c23c3eed. Also, I added the build artifact `env.js` to `.gitignore`, as of course it will vary depending on the environment it's built for.

**Comment by robinwn** [comment]
@hangler react-env  does [successive defaulting](https://github.com/andrewmclagan/react-env#arguments-and-parameters) of various .env, .env.development etc so leaving in say .env.local would be useful for anyone else (say me :-) ) checking out the code.

**Comment by hangler** [comment]
@robinwn Definitely! The actual `.env` input file is still in the repo; it's just the `env.js` file that gets built from it that's been gitignored.
